packages/utils/src/useSubscription.ts

```diff
@@ -0,0 +1,121 @@
+import { Unsubscribe as AuthUnsubscribe } from "firebase/auth";
+import { Unsubscribe as FirestoreUnsubscribe } from "firebase/firestore";
+import { Unsubscribe as DatabaseUnsubscribe } from "firebase/database";
+import { useEffect } from "react";
+import {
+  hashQueryKey,
+  QueryFunction,
+  QueryKey,
+  useQuery,
+  useQueryClient,
+  UseQueryOptions,
+  UseQueryResult,
+} from "react-query";
+
+type Unsubscribe = AuthUnsubscribe | FirestoreUnsubscribe | DatabaseUnsubscribe;
+
+const unsubscribes: Record<string, any> = {};
+const observerCount: Record<string, number> = {};
+const eventCount: Record<string, number> = {};
+
+interface CancellablePromise<T = void> extends Promise<T> {
+  cancel?: () => void;
+}
+
+type UseSubscriptionOptions<TData, TError, R> = UseQueryOptions<
+  TData,
+  TError,
+  R
+> & {
+  onlyOnce?: boolean;
+  fetchFn?: () => Promise<TData>;
+};
+
+export function useSubscription<TData, TError, R = TData>(
+  queryKey: QueryKey,
+  subscriptionKey: QueryKey,
+  subscribeFn: (cb: (data: TData | null) => Promise<void>) => Unsubscribe,
+  options?: UseSubscriptionOptions<TData, TError, R>
+): UseQueryResult<R, TError> {
+  useEffect(() => {
+    if (!options?.onlyOnce) {
+      observerCount[subscriptionHash] += 1;
+      return () => {
+        observerCount[subscriptionHash] -= 1;
+        cleanupSubscription(subscriptionHash);
+      };
+    }
+  }, []);
+
+  const hashFn = options?.queryKeyHashFn || hashQueryKey;
+  const subscriptionHash = hashFn(subscriptionKey);
+
+  if (!options?.onlyOnce) {
```

Comment:
realtime database's subscription flag value is not `onlyOnce` but `subscribe`.

It seems that refactoring is needed.